### PR TITLE
make shelves activitypub actors

### PR DESCRIPTION
**Pull request by jum-s**
shelf as actors 
- actor name is `shelf:shelfId`, 
  - prefix `shelf` is welcome to stay in order to disambiguate outbox and follow from the controller (without requesting db doc)
  - `:` needs some occasional `decodeURIComponent` maybe a `-` is more straightforward ?
- implements `actor` `outbox` and `follow` endpoints
- creates activity on `addItems` only

ways of improvements: 
- create activity on other events : such as `item:udpate:listing` `private -> public`
- dereference activity on `removeItems`
- with a more reable sluggish actor name (ie. `stableUsername:shelf:stableShefName`)

### CODE REVIEW

**Comment by maxlath** [comment]
misses webfinger implementation and tests, which should probably be done after a rebase on the new `activity-pub-entities` tip, see  https://github.com/inventaire/inventaire/pull/552#issuecomment-928019534

**Comment by jum-s** [comment]
done with 5662357

**Review by maxlath — CHANGES_REQUESTED** [review]

File: server/controllers/activitypub/lib/format_items_activities.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      type: 'Create',
+      object,
+      actor,
+      to: 'Public',
+    }
+  },
+  findFullRangeFromActivities: activitiesDocs => {
+    return {
+      since: _.min(_.map(activitiesDocs, 'object.items.since')),
+      until: _.max(_.map(activitiesDocs, 'object.items.until'))
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/format_items_activities.js:41]: i unsucessfully tried this and finally dropped it

File: server/controllers/activitypub/lib/radio_hooks.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+const debounceActivities = async shelfId => {
+  if (!debouncedActivities[shelfId]) {
+    debouncedActivities[shelfId] = _.debounce(createDebouncedActivity({ shelfId }), activitiesDebounceTime)
+  }
+  return debouncedActivities[shelfId]()
+}
+
+const _deliverEntityActivitiesFromPatch = patch => {
+  deliverEntityActivitiesFromPatch(patch)
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/radio_hooks.js]: solved 4f7e0508

File: server/controllers/activitypub/lib/radio_hooks.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+module.exports = () => {
+  radio.on('user:inventory:update', userId => {
+    if (!debouncedActivities[userId]) {
+      debouncedActivities[userId] = _.debounce(createDebouncedActivity({ userId }), activitiesDebounceTime)
+    }
+    return debouncedActivities[userId]()
+  })
+  radio.on('shelves:update', shelvesIds => {
+    Promise.all(shelvesIds.map(debounceActivities))
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/radio_hooks.js]: solved 4f7e0508

File: server/controllers/shelves/lib/shelves.js [hunk 4]
```diff
@@ -30,8 +31,10 @@ const shelves_ = module.exports = {
     const updatedShelf = Shelf.updateAttributes(shelf, newAttributes, reqUserId)
     return db.putAndReturn(updatedShelf)
   },
-  addItems: (ids, itemsIds, userId) => {
-    return updateShelvesItems('addShelves', ids, userId, itemsIds)
+  addItems: async (ids, itemsIds, userId) => {
+    const docs = await updateShelvesItems('addShelves', ids, userId, itemsIds)
+    emit('shelves:update', ids)
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/shelves/lib/shelves.js]: solved 4f7e0508

File: server/controllers/activitypub/lib/radio_hooks.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  })
+  radio.on('patch:created', _deliverEntityActivitiesFromPatch)
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/radio_hooks.js:23]: i dont, this is simply where all the hooks are, im fine leaving it here for now
Is this blocking ? i would like to merge this quickly for the moment

File: server/controllers/activitypub/lib/format_items_activities.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+const CONFIG = require('config')
+const host = CONFIG.fullPublicHost()
+const { i18n } = require('lib/emails/i18n/i18n')
+const snapshot_ = require('controllers/items/lib/snapshot/snapshot')
+const maxLinksToDisplay = 3
+
+module.exports = {
+  createItemsNote: ({ allActivitiesItems, lang, name, actor, parentLink }) => async activityDoc => {
+    const { since, until } = activityDoc.object.items
+    const publicRangeItems = allActivitiesItems.filter(itemsWithinActivityRange(since, until))
```
[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/format_items_activities.js:12]: left it like that for now but added a comment in 4f7e0508

File: server/controllers/activitypub/lib/radio_hooks.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxlath) [reply server/controllers/activitypub/lib/radio_hooks.js:23]: If speed is a problem, I can quickly split it :p I won't block on this, but it's not some trivial one liner functions that we are stacking in this file. While complex `deliverEntityActivitiesFromPatch` and its dependencies could be in a nicely isolated module, and instead here we have to scroll though the file to understand what's happening, for 0 added value

[the remark this answers is not in the document]
Comment (jum-s) [reply server/controllers/activitypub/lib/radio_hooks.js]: done in db19a0dd

**Review by maxlath — APPROVED** [review]

**Pull request merged**